### PR TITLE
[8.x] Add skip Column Statistics for php artisan schema:dump

### PR DESCRIPTION
**Pull request by roberto-butti**
#34078 

I'm creating a new PR on the 8.x branch instead of master as requested by @GrahamCampbell and @driesvints

### CODE REVIEW

**Comment by driesvints** [comment]
Did you read Taylor's comment here to check the error message? https://github.com/laravel/framework/pull/34125#issuecomment-686762944

Is that possible? Then we might not need the flag.

**Comment by roberto-butti** [comment]
> Did you read Taylor's comment here to check the error message? [#34125 (comment)](https://github.com/laravel/framework/pull/34125#issuecomment-686762944)
> 
> Is that possible? Then we might not need the flag.

I read the Taylor's comment. I was trying to identify a secure and reliable way to detect the error for not existing table. It is super easy to catch the error, is less easy to identify the right reason (mysqldump is an external process so we could rely on the exit status code). I was reading also the source code of mysqldump to see the exit status **code**. I think that we can't rely on the **message** because we have different version of mysqldump (release) and also fork (mariadb etc) (and we need to consider also the translation of the message)
I think that this it takes some times to be completed and tested so my (personal) suggestion (but you are the guru of Laravel so, the final decision is on your side) is:
- proceed with this current fix (it means that the user fails the php artisan schema:dump when has the particular configuration of mysqldump8 and mysqld5.7, and he has the opportunity to use the option --column-statistics-off to complete the dump).
- open a new issues to detect the error, and in the case we are in the error for missing column statistics table, we could autoamitically apply the statistics off.

**Comment by taylorotwell** [comment]
This whole issue is kinda wearing me down.

**Comment by taylorotwell** [comment]
You can't simply check if the error message contains `column_statistics`?

**Comment by roberto-butti** [comment]
> This whole issue is kinda wearing me down.

We want Taylor happy the week before the Laravel 8 release. So I'm preparing the PR with the check on the message error for "column_statistics" ;)

File: src/Illuminate/Database/Schema/MySqlSchemaState.php [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 class MySqlSchemaState extends SchemaState
 {
+    public $columnStatisticsOff = false;
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Schema/MySqlSchemaState.php:7]: Missing phpdoc.

File: src/Illuminate/Database/Schema/MySqlSchemaState.php [hunk 2]
```diff
@@ -78,8 +80,9 @@ public function load($path)
     protected function baseDumpCommand()
     {
         $gtidPurged = $this->connection->isMaria() ? '' : '--set-gtid-purged=OFF';
+        $columnStatisticsOption = ($this->columnStatisticsOff) ? ' --column-statistics=0' : '';
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Schema/MySqlSchemaState.php:83]: Don't need brackets.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by taylorotwell** [comment]
Isn't `appendMigrationData` going to run into the same problem? Did you test this?

**Comment by roberto-butti** [comment]
> Isn't `appendMigrationData` going to run into the same problem? Did you test this?

Exactly, appendMigrationData at the end will use mysqldump as dump command in case of Mysql. So it needs the same logic. 
But both, appendMigrationData() and dump(), are using baseDumpCommand() to produce the common mysqldump string command line with common options. The "logic" is inside  baseDumpCommand().

And yes I tested that. But my suggestion is to invite also other people involved to test because this is strictly related with envrinoments (O.S. + version of third party command -mysqldump-).
I tested this  "matrix":
- mysqldump 5.7 -> mysqld 5.7
- mysqldump 5.7 -> mysqld 8
- mysqldump 8 - > mysqld 5.7
- mysqldump 8 -> mysqld 8

**Comment by taylorotwell** [comment]
`baseDumpCommand` has no try / catch whatsoever so how will appendMigrationData use the try / catch logic to know whether to use the flag?

**Comment by roberto-butti** [comment]
> `baseDumpCommand` has no try / catch whatsoever so how will appendMigrationData use the try / catch logic to know whether to use the flag?

My precondition was (maybe is too weak): appendMigrationData is protected and called only in dump().
dump() first of all performs the try/catch when execute the process. If we are in catch condition and the error message contains column_statistics, then  I set columnStatisticsOff as true. This flag is used also by appendMigrationData (in baseDumpCommand() ).

If you want to perform the check also in  appendMigrationData let me know, I can add try catch, 
OR if you have a better solution let me know.
Thank you

**Comment by taylorotwell** [comment]
I'm sorry but this is incredibly confusing PR. You have a local variable with a very confusing name. You are mutating that to control another part of code, etc. You should just pass local variables. You don't need any local state on the class at all to accomplish this feature. I can't even really process the PR anymore because it's so difficult to understand. Feel free to take another stab at it if you feel like it but I've tried to work on it and can't even understand it anymore.

**Comment by taylorotwell** [comment]
Fixed: https://github.com/laravel/framework/commit/d67be1305bef418d9bdeb8192177202f9d705699

**Comment by maher1337** [comment]
@taylorotwell , this is not yet fixed.
`Error Output:                                                                                                                                                                                           
  ================                                                                                                                                                                                        
  mysqldump: [Warning] Using a password on the command line interface can be insecure.                                                                                                                    
  mysqldump: Couldn't execute 'SELECT COLUMN_NAME,                       JSON_EXTRACT(HISTOGRAM, '$."number-of-buckets-specified"')                FROM information_schema.COLUMN_STATISTICS              
      WHERE SCHEMA_NAME = 'EXAMPLE' AND TABLE_NAME = 'EXAMPLE';': Unknown table 'COLUMN_STATISTICS' in information_schema (1109) `

**Pull request closed without merging**